NetKet/Graph/lattice_impl.hpp

```diff
@@ -0,0 +1,430 @@
+    for (int j = 0; j < Ndim(); j++) {
+      int new_index = n[j] + lattice_neighbours_vector[i][j];
+      if (pbc_[j]) {
+        if (new_index < 0)
+          new_index = extent_[j] - 1;
+        else if (new_index >= extent_[j])
+          new_index = 0;
+        single_neighbour_vector.push_back(new_index);
+
+      } else if (new_index >= 0 and new_index < extent_[j])
+        single_neighbour_vector.push_back(new_index);
+
+      else
+        flag = false;  // if new_index refers to a border site, don't add this
+                       // site to the neighbours
+    }
+    if (flag) {
+      single_neighbour.push_back(Vector2Site(single_neighbour_vector) +
+                                 lattice_neighbours_vector[i][Ndim()] *
+                                     nlatticesites_);
+    }
+  }
+
+  return single_neighbour;
+}
+
+// Edges and AdjacencyList
+std::vector<Lattice::Edge> Lattice::BuildEdges() const {
+  std::vector<Lattice::Edge> edge_vector;
+  for (int k = 0; k < nlatticesites_; k++) {
+    for (int iatom = 0; iatom < natoms_; iatom++) {
+      std::vector<int> neighbours;
+
+      neighbours = Lattice::FindNeighbours(k, iatom);
+
+      for (std::size_t i = 0; i < neighbours.size(); i++) {
+        if (k + iatom * nlatticesites_ < neighbours[i])
+          edge_vector.push_back({{k + iatom * nlatticesites_, neighbours[i]}});
+      }
+    }
+  }
+  return edge_vector;
+};
+
+std::vector<std::vector<int>> Lattice::AdjacencyList() const noexcept {
+  std::vector<std::vector<int>> adlist(nsites_);
+  for (int k = 0; k < nsites_; k++) {
+    for (std::size_t i = 0; i < edges_.size(); i++) {
+      if (edges_[i][0] == k)
+        adlist[k].push_back(edges_[i][1]);
+      else if (edges_[i][1] == k)
+        adlist[k].push_back(edges_[i][0]);
+    }
+  }
+  return adlist;
+}
+
+// Symmetries
+std::vector<std::vector<int>> Lattice::BuildSymmetryTable() const {
+  std::vector<std::vector<int>> integers_all;
+  std::vector<std::vector<int>> result;
+  for (int k = 0; k < nlatticesites_; k++)
+    integers_all.push_back(Site2Vector(k));
+  int i = 0;
+  do {
+    std::vector<int> row;
+    int index;
+    for (int iatom = 0; iatom < natoms_; iatom++) {
+      for (int j = 0; j < nlatticesites_; j++) {
+        std::vector<int> n(ndim_);
+        index = 1;
+        for (int l = 0; l < ndim_; l++) {
+          if (pbc_[l]) {
+            n[l] = (integers_all[i][l] + integers_all[j][l]) % (extent_[l]);
+          } else {
+            n[l] = integers_all[j][l];
+            index *= extent_[l];
+          }
+        }
+        row.push_back(Vector2Site(n) + iatom * nlatticesites_);  // save rows
+      }
+    }
+    i += index;
+    result.push_back(row);
+  } while (i < nlatticesites_);
+
+  return result;
+};
+
+bool Lattice::IsBipartite() const noexcept { return is_bipartite_; }
+
+bool Lattice::IsConnected() const noexcept { return is_connected_; }
+
+// Utilities
+bool Lattice::RelativelyEqual(double a, double b,

```

Comment:
This function should probably be `static`.

Reply:
I put the function in `Utils/math_helpers.hpp` as @femtobit suggested so it should be ok without the static keyword